WIP: Token info page integration

DavideSegullo opened the pull request:
## What is the purpose of the change

this PR introduces the detail page for the tokens listed in the dapp.

jonator commented on packages/web/pages/assets/[denom].tsx:
```diff
@@ -78,28 +116,90 @@ const AssetInfoView = observer(() => {
   const { assetInfoConfig } = useAssetInfoView();
+  const { chainStore } = useStore();
+  const t = useTranslation();
+  const [favoritesList, setFavoritesList] = useLocalStorageState(
```
Would be great if we can somehow use the same favorites setting as in assets page.

DavideSegullo replied:
It's the same one, we use "favoritesList" key to store and retrieve data, like you do on assets page. I tried it and the states between pages seem to be synchronized, maybe I'm missing something?

jonator replied:
Ohh ok. Nice. Missed that.

jonator commented on packages/web/public/icons/smaller-caret-down.svg:
May be good to add this to sprite.svg as well. Size and color are better controlled there as well.

DavideSegullo replied:
Same as https://github.com/osmosis-labs/osmosis-frontend/pull/2234#discussion_r1351954267

jonator commented on packages/web/public/images/staking-apr-full.svg:
I think this is already in stage as osmo-tokens.svg

jonator replied:
OK SGTM

jonator commented on packages/web/styles/globals.css:
```diff
@@ -22,6 +22,9 @@
   ::-webkit-scrollbar-corner {
     background: rgba(0, 0, 0, 0);
   }
+  .tokendetailshadow::after {
```
Rationale for putting this in globals.css instead of using directly?

DavideSegullo replied:
To make the code less verbose, obviously you can use it directly if this is the convention you usually use maybe it's better not to change it, how about that? Do we use it directly or leave it as is?

jonator replied:
Up to you, thought I'd mention

DavideSegullo replied:
I would leave it like that, with tailwind with the use of "after" we would have a very verbose css class

jonator commented:
Also @DavideSegullo may want to update from stage and set this ENV var to get pools to load:

`NEXT_PUBLIC_TIMESERIES_DATA_URL=https://api-stage-osmosis.imperator.co/`

DavideSegullo commented on packages/stores/src/queries-external/coingecko-coin-infos/index.ts:
```diff
@@ -0,0 +1,82 @@
+  };
+};
+
+/** Queries Imperator for extrapolated APR of a given position's tick range. */
```
Yes, done here: ae0c09a68ddc31b0c70356d7c9b2d225786f2559

DavideSegullo commented on packages/stores/src/queries-external/circulating-supply/index.ts:
```diff
@@ -0,0 +1,55 @@
+type Response = {
+  amount: number;
+};
+
```
Yes, done here: 91c8bb5311b7c050fcb1489753f249d25722f1a6

DavideSegullo commented on packages/stores/src/queries-external/mcap/index.ts:
```diff
@@ -0,0 +1,35 @@
+import { ObservableQueryExternalBase } from "../base";
+import { MarketCap } from "./types";
+
+/** Queries Imperator token history data chart. */
```
Yes, done here: 388f0cddb15724ded87c1a7ccac978f9ef358afd

DavideSegullo commented on packages/web/components/your-balance/your-balance.tsx:
```diff
@@ -0,0 +1,188 @@
+  denom: string;
+}
+
+function YourBalance({ denom }: YourBalanceProps) {
```
It's already wrapped: https://github.com/osmosis-labs/osmosis-frontend/pull/2234/files/0b283da901e888ec3f9ce998980f969b60619415#diff-8f451f8c5cc352b7cdf1845cd5482b9e0c944e0d0137257161c32a070d8666ddR93 , Would you prefer that I move it to where I declare the functional component?

jonator replied:
Ohh I see. Yeah I think it would be more visible IMO

DavideSegullo commented on packages/web/components/assets/caret-down.tsx:
```diff
@@ -0,0 +1,18 @@
+
+function CaretDown({ className }: { className?: string }) {
+  return (
+    <svg
```
I saw that there are different approaches in the code to handle icons, is the sprite.svg one the convention you are starting to adopt? If yes we move the assets the

jonator replied:
Yep we prefer small icons to be placed in sprite.svg. More optionality around styling and less network requests.

DavideSegullo replied:
I'll fix that

DavideSegullo commented on packages/web/components/related-assets/related-assets.tsx:
```diff
@@ -0,0 +1,170 @@
+    assetsStore.unverifiedIbcBalances
+  );
+
+  const assetsToDisplay = new Set<string>();
+  for (const pool of memoedPools) {
+    if (assetsToDisplay.size === numberOfAssetsToDisplay) {
+      break;
+    }
+    if (pool.poolAssets.some((asset) => asset.amount.denom === tokenDenom)) {
+      assetsToDisplay.add(
+        pool.poolAssets.filter((asset) => asset.amount.denom !== tokenDenom)[0]
+          .amount.denom
+      );
+    }
+  }
```
We do not need a memo, because we are directly computing the mobx state, the result is already memoized , but we can improve the code to move the logic and avoid having it inline

DavideSegullo replied:
I also did a bit of refactoring: cba0c55456f23e9c76d4be5e1a18dd0ba2960177

jonator commented:
@DavideSegullo does it still need to be a draft? I'm going to have our QA team look at it as well

DavideSegullo commented on packages/web/components/swap-tool/index.tsx:
```diff
@@ -812,119 +953,128 @@ export const SwapTool: FunctionComponent<{
 
             <SkeletonLoader
```
I had missed this comment, the loading animation part we didn't touch, or maybe the new header causes some misalignment? What am I missing?

jonator replied:
I'm talking about the expandable details section below the inputs. When loading it appears differently to what we currently have in stage.

jonator replied:
![Screenshot 2023-10-11 at 10 11 24 AM](https://github.com/osmosis-labs/osmosis-frontend/assets/4606373/06fc9e59-c102-4bc9-a189-6fd5264acc28)

DavideSegullo replied:
We did this because in the new design there are changes to the component the bottom section when collapsed is large compared to before and shows more information

<img width="379" alt="Screenshot 2023-10-11 alle 16 37 19" src="https://github.com/osmosis-labs/osmosis-frontend/assets/17269969/028eb775-3c1c-4f54-9252-f193f4c38788">

jonator replied:
Can we leave that area as it is? I'll sync with Syed but I think that was unintentional, and changes there are not worth it IMO

DavideSegullo replied:
Yes, I'll reverse it

DavideSegullo replied:
Done here: https://github.com/osmosis-labs/osmosis-frontend/pull/2234/commits/8e793e257f6db62a2e57d9ab21747594af5fe721

JoseRFelix commented on packages/web/services/github.ts:
```diff
@@ -0,0 +1,27 @@
+import axios from "axios";
```
How about moving this to packages/web/queries, where we typically store queries to be run on the server?

DavideSegullo replied:
Yes, it sounds good!

DavideSegullo replied:
Done https://github.com/osmosis-labs/osmosis-frontend/pull/2234/commits/a87650c9959226fc71828eabaa13c50d1bd75336

JoseRFelix commented on packages/web/services/twitter.ts:
```diff
@@ -0,0 +1,112 @@
+import { createClient } from "@vercel/kv";
```
Same as above

DavideSegullo replied:
Done https://github.com/osmosis-labs/osmosis-frontend/pull/2234/commits/a87650c9959226fc71828eabaa13c50d1bd75336

JoseRFelix commented on packages/web/services/twitter.ts:
```diff
@@ -0,0 +1,112 @@
+        previewImage: mediaTweet !== undefined ? mediaTweet.url : null,
+      };
+    });
+  },
+};
+
+/**
+ * Expire time in milliseconds. (7 days)
+ *
+ * It should be enought if we wanna get tweets from 35 tokens
+ * every 7 days.
+ */
+const CACHE_EXPIRE_TIME = 1000 * 60 * 60 * 24 * 7;
+
+export const Twitter: TwitterServiceInterface = {
+  /**
+   * Returns the tweets associated with the userId passed as a parameter,
+   * if it finds them in cache it returns them if not it makes a request
+   * to Twitter and then caches the result
+   *
+   * @param userId a search string (e.g. "osmosiszone")
+   * @returns An array of tweet's objects
+   */
+  getUserTweets: async (userId: string) => {
+    const twitter = createClient({
+      url: process.env.KV_REST_API_URL!,
+      token: process.env.KV_REST_API_TOKEN!,
+    });
+
+    let cachedTweets = await twitter.get<RichTweet[]>(userId);
+
+    if (!cachedTweets) {
+      cachedTweets = await TwitterInternal.getUserTweets(userId);
+
+      twitter.set(userId, cachedTweets, { px: CACHE_EXPIRE_TIME, nx: true });
+    }
+
+    return cachedTweets;
+  },
+};
```
I suggest creating a class to streamline the underlying logic. Additionally, I recommend utilizing the `cachified` utility from our [queries](https://github.com/osmosis-labs/osmosis-frontend/blob/85b7c7372734044139ff4d09f9677d38a491e35c/packages/web/queries/complex/pools.ts#L92). It's straightforward to implement and manages all caching, even with Vercel KV using the Redis adapter."

DavideSegullo replied:
Done here: https://github.com/osmosis-labs/osmosis-frontend/pull/2234/commits/458a15d69724e86de92e0962e873612b01bcad03

JoseRFelix commented on packages/web/services/twitter.ts:
```diff
@@ -0,0 +1,112 @@
+
+import { TWITTER_API_ACCESS_TOKEN, TWITTER_API_URL } from "~/config";
+
+const twitterApi = axios.create({
+  baseURL: TWITTER_API_URL,
+  headers: {
+    Authorization: `Bearer ${TWITTER_API_ACCESS_TOKEN}`,
+  },
+});
```
I recommend using `fetch` as it's built-in and won't add to the bundle size. I've developed an [API client](https://github.com/osmosis-labs/osmosis-frontend/pull/2208/files#diff-ceb538631f37f488ab98b32f93c1cad7dd499df2fc8859f5657d2db817b4a886) in my work-in-progress PR #2208. Please consider incorporating it into this PR."

DavideSegullo replied:
Yes, I'll refactor that, I used axios because I had seen it around the codebase

DavideSegullo replied:
I saw that PR is still in draft, as soon as you implement it on stage branch, I will do a rebase and use your client, there are libraries built using only fetch that you might be interested in: https://github.com/sindresorhus/ky (maybe in the future for advanced cases)

JoseRFelix commented on packages/web/services/github.ts:
```diff
@@ -0,0 +1,27 @@
+
+import { CMS_REPOSITORY_PATH, GITHUB_URL } from "~/config";
+
+const githubApi = axios.create({
+  baseURL: GITHUB_URL,
+  headers: {
+    Accept: "application/vnd.github.v3+json",
+  },
+});
```
Same as above

syed8rahman commented:
<img width="1148" alt="Screen Shot 2023-10-16 at 10 50 20 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/966c465b-8a87-4dc7-8c65-233977379edc">
Seems to continuously crash after ~10 mins of being open

syed8rahman commented:
<img width="1145" alt="Screen Shot 2023-10-16 at 10 51 03 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/cd3160bd-31da-489c-9406-bc67f2e96691">
When condensed (probably tablet mode) swap tool shouldn't be at the bottom here. Ideally we keep price chart up top and then swap + earn are side by side

syed8rahman commented:
<img width="1327" alt="Screen Shot 2023-10-16 at 10 52 35 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/0f787f6f-b18b-4522-ae52-a98f11fd7151">
7 day 30 day 1 year are mixed up. Also need 1D. Also lets adjust to match the labels below - 1H, 
1D, 7D, 30D, 1Y
<img width="974" alt="Screen Shot 2023-10-16 at 10 53 20 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/fb35d139-a0ed-4ea6-b9bd-bd66b22d0282">

syed8rahman commented:
<img width="966" alt="Screen Shot 2023-10-16 at 10 54 41 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/4eee2cc1-03c2-4d96-bc5f-72258609a4f0">

Adjusted some colors and font selections - changed body from Subtitle to Body 2 for better readability. 
<img width="955" alt="Screen Shot 2023-10-16 at 10 55 43 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/e3e18a30-36a8-4165-aeae-a748110129f6">

syed8rahman commented:
<img width="939" alt="Screen Shot 2023-10-16 at 10 56 39 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/63cec50b-f3c6-4008-bed5-fd8d9566be23">
Dividers between tweets should also be 1pt stroke and a lighter color.

syed8rahman commented:
<img width="295" alt="Screen Shot 2023-10-16 at 11 03 27 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/ceee2d28-a7c8-4e39-898e-e425ec337f18">
Adjusted internal padding for the related assets section

syed8rahman commented:
<img width="461" alt="Screen Shot 2023-10-16 at 11 23 39 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/9400da28-95fa-47ab-8d4a-11f0e412bc6c">
This may already be a known issue but related assets are flickering between different lists every half second.

syed8rahman commented:
<img width="660" alt="Screen Shot 2023-10-16 at 11 29 07 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/bf026b7b-21aa-44dc-8754-8f7c55f4446b">
Created a smaller variant for the Earn card if condensed - let me know if this is feasible

DavideSegullo commented:
> <img alt="Screen Shot 2023-10-16 at 11 23 39 AM" width="461" src="https://user-images.githubusercontent.com/22246254/275597319-9400da28-95fa-47ab-8d4a-11f0e412bc6c.png"> This may already be a known issue but related assets are flickering between different lists every half second.

Fixed https://github.com/osmosis-labs/osmosis-frontend/pull/2234/commits/7ff5aa048347c60ac7598973d2a30e331c590d4b , I fixed a problem in the logic by which the associated assets were being taken, I don't see the refresh every half second, but let me know if that works now, thanks!

DavideSegullo commented:
> When condensed (probably tablet mode) swap tool shouldn't be at the bottom here. Ideally we keep price chart up top and then swap + earn are side by side

But if we move the swap tool to the right of the graph, we remove the space for the tweet section and the earn section, and the elements overlap, maybe I'm missing something, do you have an example to show me?

syed8rahman commented:
<img width="1170" alt="Screen Shot 2023-10-20 at 11 45 28 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/3c945178-5170-4cf5-a2fe-db50623e639a">
All the rows in the Assets page need a hover state highlight so users are aware they can be clicked into for the token page.

syed8rahman commented:
<img width="919" alt="Screen Shot 2023-10-20 at 11 46 45 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/b9e45cd7-fe78-4409-bacd-a13cf0cf5cd0">
Hover state for row items

syed8rahman commented:
<img width="786" alt="Screen Shot 2023-10-20 at 11 48 05 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/ae731da9-b60f-4a70-841f-0edd62a5297d">
The X-axis label should extend to the end - missing Friday the 20th.

syed8rahman commented:
<img width="473" alt="Screen Shot 2023-10-20 at 11 50 38 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/c40ee1f8-dcaa-4658-ac50-d5bdfaa4d058">
Swap and Buy headers are still a bit too large 
<img width="454" alt="Screen Shot 2023-10-20 at 11 50 55 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/05d1c652-8cb8-472d-bb13-f543bcacda82">

syed8rahman commented:
<img width="1263" alt="Screen Shot 2023-10-20 at 11 52 11 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/3cbc4677-06f0-439b-a795-519beb5c648e">
It seems like the swap modal is not properly populated or functional when the wallet is disconnected. should still function like it does on the main swap page when disconnected.

syed8rahman commented:
<img width="1230" alt="Screen Shot 2023-10-20 at 11 54 46 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/9b191e14-3833-45a5-85af-a43e2f7b1678">
Not sure what the difference is between live + Figma, but maybe our swap tool specs are different from the existing component. Can we make the graph height match the (collapsed) swap tool?

syed8rahman commented:
<img width="517" alt="Screen Shot 2023-10-20 at 11 56 53 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/19f82440-aef7-4d2f-9f14-4598767fbe64">
In the tablet + mobile + wrapped state, was saying we should have the swap modal immediately under the stats, just the order in which they are stacked

syed8rahman commented:
<img width="1235" alt="Screen Shot 2023-10-20 at 11 58 30 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/8408399c-8c04-41b8-a973-5e63d207e867">
Related assets are still flickering every second or so, or disappear altogether after a second. Also sometimes only appears once a wallet is connected.

syed8rahman commented:
<img width="769" alt="Screen Shot 2023-10-20 at 11 59 15 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/c6d848b6-3776-4d53-a53a-8aaef514dc37">
The x-axis labels seem wrong here in the 7 day state

syed8rahman commented:
Also y-axis is cut off. The intention actually was to not have a y-axis but rather a hover state when mousing over the graph to show the price, similar to coinbase. also without the horizontal lines on the chart, just to keep things nice and clean. 
<img width="774" alt="Screen Shot 2023-10-20 at 11 59 56 AM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/6137cd98-d812-44dd-b7e4-ed30cfb01ca4">
<img width="802" alt="Screen Shot 2023-10-20 at 12 00 44 PM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/5ad89085-6d5f-45e5-9654-0d7d66a024de">
<img width="879" alt="Screen Shot 2023-10-20 at 12 01 28 PM" src="https://github.com/osmosis-labs/osmosis-frontend/assets/22246254/14bb4a86-a9c3-4cf0-8901-f3461031e6c0">

DavideSegullo commented:
> <img alt="Screen Shot 2023-10-20 at 11 45 28 AM" width="1170" src="https://user-images.githubusercontent.com/22246254/277012938-3c945178-5170-4cf5-a2fe-db50623e639a.png"> All the rows in the Assets page need a hover state highlight so users are aware they can be clicked into for the token page.

Done https://github.com/osmosis-labs/osmosis-frontend/pull/2234/commits/1bd146547077564c8c924f5192ec4295fec73088 , there was already a style available in the table, now the whole row is clickable
